brainiak/utils/fmrisim.py

```diff
@@ -2046,38 +2247,345 @@ def _noise_dict_update(noise_dict):
+            ending with sigma (e.g. motion sigma) are mixed together in
+            _generate_temporal_noise. These values describe the proportion of
+            mixing of these elements. However critically, SFNR is the
+            parameter that describes how much noise these components contribute
+            to the brain. If you set the noise dict to matched then it will
+            fit the parameters to match the participant as best as possible.
+
+        fit_thresh : float
+            What proportion of the target parameter value is sufficient
+            error to warrant finishing fit search.
+
+        fit_delta : float
+            How much are the parameters attenuated during the fitting process,
+            in terms of the proportion of difference between the target
+            parameter and the actual parameter
+
+        iterations : list, int
+            The first element is how many steps of fitting the SFNR and SNR
+            values will be performed. Usually converges after < 5. The
+            second element is the number of iterations for the AR fitting.
+            This is much more time consuming (has to make a new timecourse
+            on each iteration) so be careful about setting this appropriately.
+
+        Returns
+        -------
+
+        noise : multidimensional array, float
+            Generates the noise volume given these parameters
+
+    """
+
+    # Pull out the
+    dim_tr = noise.shape
+    dim = dim_tr[0:3]
+    base = template * noise_dict['max_activity']
+    base = base.reshape(dim[0], dim[1], dim[2], 1)
+    mean_signal = (base[mask > 0]).mean()
+
+    # Iterate through different parameters to fit SNR and SFNR
+    temp_sd_orig = np.copy(temporal_sd)
+
+    # Make a copy of the dictionary so it can be modified
+    new_nd = copy.deepcopy(noise_dict)
+
+    # What SFNR do you want
+    target_sfnr = noise_dict['sfnr']
+
+    # What AR do you want?
+    target_ar = noise_dict['auto_reg_rho'][0]
+
+    # Iterate through different MA parameters to fit AR
+    for iteration in list(range(iterations)):
+
+        # If there are iterations left to perform then recalculate the
+        # metrics and try again
+
+        # Calculate the new SFNR
+        new_sfnr = _calc_sfnr(noise, mask)
+
+        # Calculate the AR
+        new_ar, _ = _calc_ARMA_noise(noise,
+                                     mask,
+                                     len(noise_dict['auto_reg_rho']),
+                                     len(noise_dict['ma_rho']),
+                                     )
+
+        # Calculate the difference between the real and simulated data
+        sfnr_diff = abs(new_sfnr - target_sfnr) / target_sfnr
+
+        # Calculate the difference in the first AR component
+        ar_diff = new_ar[0] - target_ar
+
+        # If the SFNR and AR is sufficiently close then break the loop
+        if (abs(ar_diff) / target_ar) < fit_thresh and sfnr_diff < fit_thresh:
```

Comment:
I am not sure I understand the suggestion: do you mean to use more components (e.g. AR2) or do you mean some other parameters from an ARMA process? In general I have found it hard to handle the simulation when the AR is greater than 1 component

Reply:
Yeah so to answer the last question, I have found it hard to get all parameters to converge simultaneously. Any help on that would be appreciated. For now, I have found it sufficient to fit the SNR/spatial noise first, and then the temporal noise second since the latter doesn't affect the former much. In the case of temporal fitting, I am fitting all of the temporal noise properties that we are trying to fit. But it is an interesting suggestion to have an omnibus threshold rather than specific ones for each parameter, although I feel like that is ultimately less conservative right (e.g. some parameters may be fit well, whereas others are fit poorly but on average they are below threshold)

Reply:
Ah I should have written to check the maximum instead of the minimum absolute fitting errors for all parameters to be smaller than the threshold.
Have you tried to monitor for a few iterations whether the parameters that does not converge just jump up and down around the desired value, or keep being above (or below) the desired value?
I think it may be worth correcting the issues related to spatial_sd and temporal_sd in another comment to see if it converges better. I am worried they may be related.